File: Raven.Client.Lightweight/Connection/RavenTransactionAccessor.cs
//-----------------------------------------------------------------------
// <copyright file="RavenTransactionAccessor.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Transactions;
using Raven.Abstractions.Extensions;
using TransactionInformation = Raven.Abstractions.Data.TransactionInformation;

namespace Raven.Client.Connection
{
	/// <summary>
	/// Provide access to the current transaction 
	/// </summary>
	public static class RavenTransactionAccessor
	{
		[ThreadStatic]
		private static Stack<TransactionInformation> currentRavenTransactions;

		private static Stack<TransactionInformation> CurrentRavenTransactions
		{
			get
			{
				if(currentRavenTransactions == null)
					currentRavenTransactions =  new Stack<TransactionInformation>();
				return currentRavenTransactions;
			}
		}

		[ThreadStatic]
		private static bool supressExplicitRavenTransaction = false;

		/// <summary>
		/// Starts a transaction
		/// </summary>
		/// <returns></returns>
		public static IDisposable StartTransaction()
		{
			return StartTransaction(TimeSpan.FromMinutes(1));
		}

		/// <summary>
		/// Starts a transaction with the specified timeout
		/// </summary>
		/// <param name="timeout">The timeout.</param>
		/// <returns></returns>
		public static IDisposable StartTransaction(TimeSpan timeout)
		{
			CurrentRavenTransactions.Push(new TransactionInformation
			{
				Id = Guid.NewGuid().ToString(),
				Timeout = timeout
			});
			return new DisposableAction(() => CurrentRavenTransactions.Pop());
		}
		
		public static TimeSpan? DefaultTimeout { get; set; }


		internal static IDisposable SupressExplicitRavenTransaction()
		{
			supressExplicitRavenTransaction = true;
			return new DisposableAction(() =>
			{
				supressExplicitRavenTransaction = false;
			});
		}
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/client/connection/RavenTransactionAccessor.java
package raven.client.connection;

import java.util.Stack;
import java.util.UUID;

import raven.abstractions.data.TransactionInformation;

public class RavenTransactionAccessor {
  private RavenTransactionAccessor() {
    // empty by design
  }

  private static ThreadLocal<Stack<TransactionInformation>> currentRavenTransactions = new ThreadLocal<>();
  private static ThreadLocal<Boolean> supressExplicitRavenTransaction = new ThreadLocal<>();

  private static Stack<TransactionInformation> getCurrentRavenTransactions() {
    if (currentRavenTransactions.get() == null) {
      currentRavenTransactions.set(new Stack<TransactionInformation>());
    }
    return currentRavenTransactions.get();
  }

  private static boolean isSupressExplicitRavenTransaction() {
    if (supressExplicitRavenTransaction.get() == null) {
      supressExplicitRavenTransaction.set(false);
    }
    return supressExplicitRavenTransaction.get();
  }

  /**
   * Starts a transaction
   * @return
   */
  public static AutoCloseable startTransaction() {
    return startTransaction((long)60 * 1000);
  }

  /**
   * Starts a transaction with the specified timeout
   * @param timeout timeout in milis
   * @return
   */
  public static AutoCloseable startTransaction(long timeoutInMilis)
  {
    TransactionInformation transInfo = new TransactionInformation();
    transInfo.setId(UUID.randomUUID().toString());
    transInfo.setTimeout(timeoutInMilis);
    getCurrentRavenTransactions().push(transInfo);
    return new AutoCloseable() {

      @Override
      public void close() throws Exception {
        getCurrentRavenTransactions().pop();
      }
    };
  }

  private static long defaultTimeout;

  public static long getDefaultTimeout() {
    return defaultTimeout;
  }

  public static void setDefaultTimeout(long defaultTimeout) {
    RavenTransactionAccessor.defaultTimeout = defaultTimeout;
  }

  public static TransactionInformation getTransactionInformation() {
    Stack<TransactionInformation> stack = currentRavenTransactions.get();
    if (!stack.isEmpty() && isSupressExplicitRavenTransaction() == false) {
      return currentRavenTransactions.get().peek();
    }
    throw new IllegalStateException("Unable to find transaction");
  }

  public static AutoCloseable supressExplicitRavenTransaction() {
    supressExplicitRavenTransaction.set(true);
    return new AutoCloseable() {

      @Override
      public void close() throws Exception {
        supressExplicitRavenTransaction.set(false);
      }
    };
  }

}
